File: training_data/reflection-tuning/day17_part1_2018_failed.py
def parse_input(lines):
    clay = set()
    min_y, max_y = float('inf'), 0
    for line in lines:
        parts = line.split(', ')
        if parts[0].startswith('x'):
            x = int(parts[0].split('=')[1])
            y_range = list(map(int, parts[1].split('=')[1].split('..')))
            for y in range(y_range[0], y_range[1] + 1):
                clay.add((x, y))
                min_y = min(min_y, y)
                max_y = max(max_y, y)
        else:
            y = int(parts[0].split('=')[1])
            x_range = list(map(int, parts[1].split('=')[1].split('..')))
            for x in range(x_range[0], x_range[1] + 1):
                clay.add((x, y))
                min_y = min(min_y, y)
                max_y = max(max_y, y)
    return clay, min_y, max_y

def flow_water(clay, min_y, max_y):
    water = set()
    flowing = set()
    
    def pour(x, y):
        if y > max_y:
            return False
        if (x, y) in clay or (x, y) in water:
            return True
        if (x, y) in flowing:
            return False
        
        flowing.add((x, y))
        
        below = pour(x, y + 1)
        if not below:
            return False
        
        left = pour(x - 1, y)
        right = pour(x + 1, y)
        
        if left and right:
            water.add((x, y))
            return True
        return False
    
    pour(500, 0)
    return water, flowing

def count_water_tiles(clay, min_y, max_y):
    water, flowing = flow_water(clay, min_y, max_y)
    return sum(1 for x, y in (water | flowing) if min_y <= y <= max_y)

def solve(input_data):
    clay, min_y, max_y = parse_input(input_data.strip().split('\n'))
    return count_water_tiles(clay, min_y, max_y)

# Example usage:
# input_data = """x=495, y=2..7
# y=7, x=495..501
# x=501, y=3..7
# x=498, y=2..4
# x=506, y=1..2
# x=498, y=10..13
# x=504, y=10..13
# y=13, x=498..504"""
# print(solve(input_data))

File: training_data/reflection-tuning/day17_part1_2018_failed.js
function countWaterTiles(input) {
  const clay = new Set();
  let minY = Infinity, maxY = -Infinity;

  // Parse input
  input.split('\n').forEach(line => {
    const [a, b] = line.split(', ');
    const [axis1, value1] = a.split('=');
    const [axis2, range] = b.split('=');
    const [start, end] = range.split('..').map(Number);

    for (let i = start; i <= end; i++) {
      const x = axis1 === 'x' ? Number(value1) : i;
      const y = axis1 === 'y' ? Number(value1) : i;
      clay.add(`${x},${y}`);
      minY = Math.min(minY, y);
      maxY = Math.max(maxY, y);
    }
  });

  const water = new Set();
  const flowing = new Set();

  function pour(x, y) {
    if (y > maxY) return false;
    if (water.has(`${x},${y}`) || clay.has(`${x},${y}`)) return true;
    if (flowing.has(`${x},${y}`)) return false;

    flowing.add(`${x},${y}`);

    if (!pour(x, y + 1)) return false;

    let left = pour(x - 1, y);
    let right = pour(x + 1, y);

    if (left && right) {
      for (let i = x; !clay.has(`${i},${y}`); i--) water.add(`${i},${y}`);
      for (let i = x + 1; !clay.has(`${i},${y}`); i++) water.add(`${i},${y}`);
      return true;
    }

    return false;
  }

  pour(500, 0);

  let count = 0;
  for (let y = minY; y <= maxY; y++) {
    for (let x = 0; x < 1000; x++) {
      if (water.has(`${x},${y}`) || flowing.has(`${x},${y}`)) count++;
    }
  }

  return count;
}

// Example usage:
const input = `x=495, y=2..7
y=7, x=495..501
x=501, y=3..7
x=498, y=2..4
x=506, y=1..2
x=498, y=10..13
x=504, y=10..13
y=13, x=498..504`;

console.log(countWaterTiles(input));
